// @webank/eslint-config-webank/rules/node.js
// https://github.com/mysticatea/eslint-plugin-node
module.exports = {
    env: {
        node: true
    },
    rules: {
        // enforce return after a callback
        // callback 之后必须立即 return
        'node/callback-return': 'off',

        // require all requires be top-level
        // require 必须在全局作用域下
        'node/global-require': 'error',

        // enforces error handling in callbacks (node environment)
        'node/handle-callback-err': 'off',

        // disallow mixing regular variable and require declarations
        'node/no-mixed-requires': ['off', false],

        // disallow use of new operator with the require function
        // 禁止直接 new require('foo');
        'node/no-new-require': 'error',

        // disallow string concatenation with __dirname and __filename
        // 禁止对 __dirname 或 __filename 使用字符串连接
        'node/no-path-concat': 'error',

        // disallow use of process.env
        'node/no-process-env': 'off',

        // disallow process.exit()
        'node/no-process-exit': 'off',

        // disallow use of synchronous methods (off by default)
        'node/no-sync': 'off',
        'node/no-unsupported-features/es-syntax': ['error', {
            version: '>=8.0.0',
            ignores: ['modules']
        }],
        'node/no-unpublished-import': ['off', {
            allowModules: [],
            convertPath: null,
            tryExtensions: ['.js', '.json', '.node']
        }]
    }
};
