// @webank/eslint-config-webank/rules/style.js
module.exports = {
    rules: {
    // enforce line breaks after opening and before closing array brackets
    // https://eslint.org/docs/rules/array-bracket-newline
    // TODO: enable? semver-major
        'array-bracket-newline': ['off', 'consistent'], // object option alternative: { multiline: true, minItems: 3 }

        // enforce line breaks between array elements
        // https://eslint.org/docs/rules/array-element-newline
        // TODO: enable? semver-major
        'array-element-newline': ['off', { multiline: true, minItems: 3 }],

        // enforce spacing inside array brackets
        // @fixable 数组的括号内的前后禁止有空格
        'array-bracket-spacing': ['error', 'never'],

        // enforce spacing inside single-line blocks
        // https://eslint.org/docs/rules/block-spacing
        // @fixable 代码块如果在一行内，那么打括号内的首尾必须有空格
        'block-spacing': ['error', 'always'],

        // enforce one true brace style
        // 强制大括号风格一致
        'brace-style': ['error', '1tbs', { allowSingleLine: true }],

        // require camel case names
        // TODO: semver-major (eslint 5): add ignoreDestructuring: false option
        // 变量名必须是 camelcase 风格的
        camelcase: ['error', {
            ignoreDestructuring: true,
            ignoreImports: true,
            properties: 'never'
        }],

        // enforce or disallow capitalization of the first letter of a comment
        // https://eslint.org/docs/rules/capitalized-comments
        'capitalized-comments': ['off', 'never', {
            line: {
                ignorePattern: '.*',
                ignoreInlineComments: true,
                ignoreConsecutiveComments: true
            },
            block: {
                ignorePattern: '.*',
                ignoreInlineComments: true,
                ignoreConsecutiveComments: true
            }
        }],
        // @fixable 对象的最后一个属性末尾必须有逗号
        // @off 没有必要
        'comma-dangle': ['error', 'never'],

        // enforce spacing before and after comma
        // @fixable 逗号前禁止有空格，逗号后必须有空格
        'comma-spacing': ['error', { before: false, after: true }],

        // enforce one true comma style
        // @fixable 禁止在首行写逗号
        'comma-style': ['error', 'last', {
            exceptions: {
                ArrayExpression: false,
                ArrayPattern: false,
                ArrowFunctionExpression: false,
                CallExpression: false,
                FunctionDeclaration: false,
                FunctionExpression: false,
                ImportDeclaration: false,
                ObjectExpression: false,
                ObjectPattern: false,
                VariableDeclaration: false,
                NewExpression: false
            }
        }],


        // disallow padding inside computed properties
        // @fixable 用作对象的计算属性时，中括号内的首尾禁止有空格
        'computed-property-spacing': ['error', 'never'],

        // enforces consistent naming when capturing the current execution context
        'consistent-this': 'off',

        // enforce newline at the end of file, with no multiple empty lines
        // @fixable 文件最后一行必须有一个空行
        'eol-last': ['error', 'always'],

        // enforce spacing between functions and their invocations
        // https://eslint.org/docs/rules/func-call-
        // @fixable 函数名和执行它的括号之间禁止有空格
        'func-call-spacing': ['error', 'never'],

        // requires function names to match the name of the variable or property to which they are
        // assigned
        // https://eslint.org/docs/rules/func-name-matching
        // TODO: semver-major (eslint 5): add considerPropertyDescriptor: true
        'func-name-matching': ['off', 'always', {
            includeCommonJSModuleExports: false
        }],

        // require function expressions to have a name
        // https://eslint.org/docs/rules/func-names
        'func-names': 'off',

        // enforces use of function declarations or expressions
        // https://eslint.org/docs/rules/func-style
        // TODO: enable
        'func-style': ['off', 'expression'],

        // enforce consistent line breaks inside function parentheses
        // https://eslint.org/docs/rules/function-paren-newline
        // 在函数参数或参数的括号内强制执行一致的换行符。
        'function-paren-newline': ['error', 'consistent'],

        // Blacklist certain identifiers to prevent them being used
        // https://eslint.org/docs/rules/id-blacklist
        'id-blacklist': 'off',

        // this option enforces minimum and maximum identifier lengths
        // (variable names, property names etc.)
        'id-length': 'off',

        // require identifiers to match the provided regular expression
        'id-match': 'off',

        // Enforce the location of arrow function bodies with implicit returns
        // https://eslint.org/docs/rules/implicit-arrow-linebreak
        // 为包含隐式返回的箭头函数强制实施一致的位置。
        'implicit-arrow-linebreak': ['error', 'beside'],

        // this option sets a specific tab width for your code
        // https://eslint.org/docs/rules/indent
        // @fixable 一个缩紧必须使用4个空格代替
        indent: ['error', 4, {
            SwitchCase: 1,
            VariableDeclarator: 1,
            outerIIFEBody: 1,
            // MemberExpression: null,
            FunctionDeclaration: {
                parameters: 1,
                body: 1
            },
            FunctionExpression: {
                parameters: 1,
                body: 1
            },
            CallExpression: {
                arguments: 1
            },
            ArrayExpression: 1,
            ObjectExpression: 1,
            ImportDeclaration: 1,
            flatTernaryExpressions: true,
            // list derived from https://github.com/benjamn/ast-types/blob/HEAD/def/jsx.js
            ignoredNodes: ['JSXElement', 'JSXElement > *', 'JSXAttribute', 'JSXIdentifier', 'JSXNamespacedName', 'JSXMemberExpression', 'JSXSpreadAttribute', 'JSXExpressionContainer', 'JSXOpeningElement', 'JSXClosingElement', 'JSXText', 'JSXEmptyExpression', 'JSXSpreadChild'],
            ignoreComments: false
        }],

        // specify whether double or single quotes should be used in JSX attributes
        // https://eslint.org/docs/rules/jsx-quotes
        'jsx-quotes': ['off', 'prefer-double'],

        // enforces spacing between keys and values in object literal properties
        // @fixable 对象字面量中冒号前面禁止有空格，后面必须有空格
        'key-spacing': ['error', { beforeColon: false, afterColon: true }],

        // require a space before & after certain keywords
        // @fixable 关键字前后必须有空格
        'keyword-spacing': ['error', {
            before: true,
            after: true,
            overrides: {
                return: { after: true },
                throw: { after: true },
                case: { after: true }
            }
        }],

        // enforce position of line comments
        // https://eslint.org/docs/rules/line-comment-position
        // TODO: enable?
        'line-comment-position': ['off', {
            position: 'above',
            ignorePattern: '',
            applyDefaultPatterns: true
        }],

        // disallow mixed 'LF' and 'CRLF' as linebreaks
        // https://eslint.org/docs/rules/linebreak-style
        // @fixable 限制换行符为 LF 或 CRLF
        'linebreak-style': ['error', 'unix'],

        // require or disallow an empty line between class members
        // https://eslint.org/docs/rules/lines-between-class-members
        // 在 class 的 members 之间添加换行
        'lines-between-class-members': ['error', 'always', { exceptAfterSingleLine: false }],

        // enforces empty lines around comments
        'lines-around-comment': 'off',

        // specify the maximum depth that blocks can be nested
        'max-depth': ['off', 4],

        // specify the maximum length of a line in your program
        // https://eslint.org/docs/rules/max-len
        'max-len': ['off', 160, 4, {
            ignoreUrls: true,
            ignoreComments: false,
            ignoreRegExpLiterals: true,
            ignoreStrings: true,
            ignoreTemplateLiterals: true
        }],
        // specify the max number of lines in a file
        // https://eslint.org/docs/rules/max-lines
        'max-lines': ['off', {
            max: 300,
            skipBlankLines: true,
            skipComments: true
        }],

        // enforce a maximum function length
        // https://eslint.org/docs/rules/max-lines-per-function
        'max-lines-per-function': ['off', {
            max: 50,
            skipBlankLines: true,
            skipComments: true,
            IIFEs: true
        }],

        // specify the maximum depth callbacks can be nested
        'max-nested-callbacks': 'off',

        // limits the number of parameters that can be used in the function declaration.
        'max-params': ['off', 3],

        // specify the maximum number of statement allowed in a function
        'max-statements': ['off', 10],

        // restrict the number of statements per line
        // https://eslint.org/docs/rules/max-statements-per-line
        'max-statements-per-line': ['off', { max: 1 }],

        // enforce a particular style for multiline comments
        // https://eslint.org/docs/rules/multiline-comment-style
        'multiline-comment-style': ['off', 'starred-block'],

        // require multiline ternary
        // https://eslint.org/docs/rules/multiline-ternary
        // TODO: enable?
        'multiline-ternary': ['off', 'never'],

        // require a capital letter for constructors
        // new 后面的类名必须首字母大写
        'new-cap': ['error', {
            newIsCap: true,
            newIsCapExceptions: [],
            capIsNew: false,
            capIsNewExceptions: ['Immutable.Map', 'Immutable.Set', 'Immutable.List']
        }],

        // disallow the omission of parentheses when invoking a constructor with no arguments
        // https://eslint.org/docs/rules/new-parens
        // @fixable new 后面的类必须有小括号
        'new-parens': 'error',

        // allow/disallow an empty newline after var statement
        'newline-after-var': 'off',

        // https://eslint.org/docs/rules/newline-before-return
        'newline-before-return': 'off',

        // enforces new line after each method call in the chain to make it
        // more readable and easy to maintain
        // https://eslint.org/docs/rules/newline-per-chained-call
        'newline-per-chained-call': ['error', { ignoreChainWithDepth: 4 }],

        // disallow use of the Array constructor
        // 禁止使用 Array 构造函数
        'no-array-constructor': 'error',

        // disallow use of bitwise operators
        // https://eslint.org/docs/rules/no-bitwise
        'no-bitwise': 'off',

        // disallow use of the continue statement
        // https://eslint.org/docs/rules/no-continue
        'no-continue': 'off',

        // disallow comments inline after code
        'no-inline-comments': 'off',

        // disallow if as the only statement in an else block
        // https://eslint.org/docs/rules/no-lonely-if
        'no-lonely-if': 'error',

        // disallow un-paren'd mixes of different operators
        // https://eslint.org/docs/rules/no-mixed-operators
        'no-mixed-operators': ['error', {
            // the list of arthmetic groups disallows mixing `%` and `**`
            // with other arithmetic operators.
            groups: [
                ['%', '**'],
                ['%', '+'],
                ['%', '-'],
                ['%', '*'],
                ['%', '/'],
                ['**', '+'],
                ['**', '-'],
                ['**', '*'],
                ['**', '/'],
                ['&', '|', '^', '~', '<<', '>>', '>>>'],
                ['==', '!=', '===', '!==', '>', '>=', '<', '<='],
                ['&&', '||'],
                ['in', 'instanceof']
            ],
            allowSamePrecedence: false
        }],

        // disallow mixed spaces and tabs for indentation
        // 禁止混用空格和 tab
        'no-mixed-spaces-and-tabs': 'error',

        // disallow use of chained assignment expressions
        // https://eslint.org/docs/rules/no-multi-assign
        // 禁止连续赋值，如 a = b = c = 4;
        'no-multi-assign': ['error'],

        // disallow multiple empty lines and only one newline at the end
        // @fixable 禁止出现超过两行的连续空行
        'no-multiple-empty-lines': ['error', { max: 2, maxEOF: 0 }],

        // disallow negated conditions
        // https://eslint.org/docs/rules/no-negated-condition
        'no-negated-condition': 'off',

        // disallow nested ternary expressions
        // 不允许嵌套的三元表达式。
        'no-nested-ternary': 'off',

        // disallow use of the Object constructor
        // 禁止直接 new Object
        'no-new-object': 'error',

        // disallow use of unary operators, ++ and --
        // https://eslint.org/docs/rules/no-plusplus
        'no-plusplus': 'off',

        // disallow certain syntax forms
        // https://eslint.org/docs/rules/no-restricted-syntax
        'no-restricted-syntax': [
            'error',
            {
                selector: 'ForInStatement',
                message: 'for..in loops iterate over the entire prototype chain, which is virtually never what you want. Use Object.{keys,values,entries}, and iterate over the resulting array.'
            },
            {
                selector: 'ForOfStatement',
                message: 'iterators/generators require regenerator-runtime, which is too heavyweight for this guide to allow them. Separately, loops should be avoided in favor of array iterations.'
            },
            {
                selector: 'LabeledStatement',
                message: 'Labels are a form of GOTO; using them makes code confusing and hard to maintain and understand.'
            },
            {
                selector: 'WithStatement',
                message: '`with` is disallowed in strict mode because it makes code impossible to predict and optimize.'
            }
        ],

        // disallow tab characters entirely
        'no-tabs': 'error',

        // disallow the use of ternary operators
        // 不允许三元运算符
        'no-ternary': 'off',

        // disallow trailing whitespace at the end of lines
        // 不允许在行尾添加尾随空白
        'no-trailing-spaces': ['error', {
            skipBlankLines: false,
            ignoreComments: false
        }],

        // disallow dangling underscores in identifiers
        // https://eslint.org/docs/rules/no-underscore-dangle
        // 是否允许下划线
        'no-underscore-dangle': ['off'],

        // disallow the use of Boolean literals in conditional expressions
        // also, prefer `a || b` over `a ? a : b`
        // https://eslint.org/docs/rules/no-unneeded-ternary
        'no-unneeded-ternary': ['error', { defaultAssignment: false }],

        // disallow whitespace before properties
        // https://eslint.org/docs/rules/no-whitespace-before-property
        // @fixable 禁止属性前面有空格，比如 foo. bar()
        'no-whitespace-before-property': 'error',

        // enforce the location of single-line statements
        // https://eslint.org/docs/rules/nonblock-statement-body-position
        // @fixable 禁止 if 后面不加大括号直接写两行代码
        'nonblock-statement-body-position': ['error', 'beside', { overrides: {} }],

        // require padding inside curly braces
        // @fixable 对象字面量只有一行时，大括号内的首尾必须有空格
        'object-curly-spacing': ['error', 'always'],

        // enforce line breaks between braces
        // https://eslint.org/docs/rules/object-curly-newline
        // @fixable 大括号内的首尾必须有换行
        'object-curly-newline': ['error', {
            ObjectExpression: { minProperties: 4, multiline: true, consistent: true },
            ObjectPattern: { minProperties: 4, multiline: true, consistent: true },
            ImportDeclaration: { minProperties: 4, multiline: true, consistent: true },
            ExportDeclaration: { minProperties: 4, multiline: true, consistent: true }
        }],

        // enforce "same line" or "multiple line" on object properties.
        // https://eslint.org/docs/rules/object-property-newline
        // 对象字面量内的属性每行必须只有一个
        'object-property-newline': ['error', {
            allowAllPropertiesOnSameLine: true
        }],

        // allow just one var statement per function
        // 禁止变量声明时，用逗号一个声明多个
        'one-var': ['error', 'never'],

        // require a newline around variable declaration
        // https://eslint.org/docs/rules/one-var-declaration-per-line
        // @fixable 变量声明必须每行一个
        'one-var-declaration-per-line': ['error', 'always'],

        // require assignment operator shorthand where possible or prohibit it entirely
        // https://eslint.org/docs/rules/operator-assignment
        'operator-assignment': ['off', 'always'],

        // Requires operator at the beginning of the line in multiline statements
        // https://eslint.org/docs/rules/operator-linebreak
        // 运算符，换行应该在行首
        'operator-linebreak': ['error', 'before', { overrides: { '=': 'none' } }],

        // disallow padding within blocks
        // TODO, semver-major: uncomment option
        // @fixable 代码块首尾必须要空行
        'padded-blocks': ['error', {
            blocks: 'never',
            classes: 'never',
            switches: 'never'
            // allowSingleLineBlocks: true,
        }],

        // Require or disallow padding lines between statements
        // https://eslint.org/docs/rules/padding-line-between-statements
        'padding-line-between-statements': 'off',

        // Prefer use of an object spread over Object.assign
        // https://eslint.org/docs/rules/prefer-object-spread
        // TODO: semver-major (eslint 5): enable
        'prefer-object-spread': 'off',

        // require quotes around object literal property names
        // https://eslint.org/docs/rules/quote-props.html
        // @fixable 对象字面量的键名禁止用引号阔起来
        'quote-props': ['error', 'as-needed', { keywords: false, unnecessary: true, numbers: false }],

        // specify whether double or single quotes should be used
        // @fixable 必须使用单引号，禁止使用双引号
        quotes: ['error', 'single', { avoidEscape: true }],

        // do not require jsdoc
        // https://eslint.org/docs/rules/require-jsdoc
        'require-jsdoc': 'off',

        // require or disallow use of semicolons instead of ASI
        // @fixable 结尾必须有分号
        semi: ['error', 'always'],

        // enforce spacing before and after semicolons
        // @fixable 一行有多个语句时，分号前面禁止有空格，分号后面必须有空格
        'semi-spacing': ['error', { before: false, after: true }],

        // Enforce location of semicolons
        // https://eslint.org/docs/rules/semi-style
        // @fixable 分号必须写在行尾，禁止写在行首
        'semi-style': ['error', 'last'],

        // requires object keys to be sorted
        'sort-keys': ['off', 'asc', { caseSensitive: false, natural: true }],

        // sort variables within the same declaration block
        'sort-vars': 'off',

        // require or disallow space before blocks
        // @fixable if, function 等的大括号之前必须要有空格，比如 if (a) {
        'space-before-blocks': 'error',

        // require or disallow space before function opening parenthesis
        // https://eslint.org/docs/rules/space-before-function-paren
        // @fixable function 的小括号之前必须要有空格
        'space-before-function-paren': ['error', {
            anonymous: 'always',
            named: 'never',
            asyncArrow: 'always'
        }],

        // require or disallow spaces inside parentheses
        // @fixable 小括号内的首尾禁止有空格
        'space-in-parens': ['error', 'never'],

        // require spaces around operators
        // @fixable 操作符左右必须有空格，比如 let sum = 1 + 2;
        'space-infix-ops': 'error',

        // Require or disallow spaces before/after unary operators
        // https://eslint.org/docs/rules/space-unary-ops
        // @fixable new, typeof 等后面必须有空格，++, -- 等禁止有空格，比如：
        // let foo = new Person();
        // bar = bar++;
        'space-unary-ops': ['error', {
            words: true,
            nonwords: false,
            overrides: {
            }
        }],

        // require or disallow a space immediately following the // or /* in a comment
        // https://eslint.org/docs/rules/spaced-comment
        // @fixable 注释的斜线或 * 后必须有空格
        'spaced-comment': ['error', 'always', {
            line: {
                exceptions: ['-', '+'],
                markers: ['=', '!'] // space here to support sprockets directives
            },
            block: {
                exceptions: ['-', '+'],
                markers: ['=', '!'], // space here to support sprockets directives
                balanced: true
            }
        }],

        // Enforce spacing around colons of switch statements
        // https://eslint.org/docs/rules/switch-colon-spacing
        // @fixable case 的冒号前禁止有空格，冒号后必须有空格
        'switch-colon-spacing': ['error', { after: true, before: false }],

        // Require or disallow spacing between template tags and their literals
        // https://eslint.org/docs/rules/template-tag-spacing
        // @fixable 模版字符串的 tag 之后禁止有空格，比如 tag`Hello World
        'template-tag-spacing': ['error', 'never'],

        // require or disallow the Unicode Byte Order Mark
        // https://eslint.org/docs/rules/unicode-bom
        // @fixable 文件开头禁止有 BOM
        'unicode-bom': ['error', 'never'],

        // require regex literals to be wrapped in parentheses
        'wrap-regex': 'off'
    }
};
