// @fesjs/compiler/lib/service/utils/pluginUtils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getPluginsOrPresets = getPluginsOrPresets;
exports.isPluginOrPreset = isPluginOrPreset;
exports.isValidPlugin = isValidPlugin;
exports.pathToObj = pathToObj;
exports.resolvePlugins = resolvePlugins;
exports.resolvePresets = resolvePresets;

var _path = require("path");

var _utils = require("@fesjs/utils");

var _enums = require("../enums");

const RE = {
  [_enums.PluginType.plugin]: /^(@fesjs\/|@webank\/fes-|fes-)plugin-/,
  [_enums.PluginType.preset]: /^(@fesjs\/|@webank\/fes-|fes-)preset-/
};

function isPluginOrPreset(type, name) {
  const hasScope = name.charAt(0) === '@';
  const re = RE[type];

  if (hasScope) {
    return re.test(name.split('/')[1]) || re.test(name);
  }

  return re.test(name);
}

function getPluginsOrPresets(type, opts) {
  const upperCaseType = type.toUpperCase();
  return [// dependencies
  // opts
  ...(opts[type === _enums.PluginType.preset ? 'presets' : 'plugins'] || []), // env
  ...(process.env[`FES_${upperCaseType}S`] || '').split(',').filter(Boolean), ...Object.keys(opts.pkg.devDependencies || {}).concat(Object.keys(opts.pkg.dependencies || {})).filter(isPluginOrPreset.bind(null, type)), // user config
  ...(opts[type === _enums.PluginType.preset ? 'userConfigPresets' : 'userConfigPlugins'] || [])].map(path => _utils.resolve.sync(path, {
    basedir: opts.cwd,
    extensions: ['.js', '.ts']
  }));
} // e.g.
// initial-state -> initialState
// webpack.css-loader -> webpack.cssLoader


function nameToKey(name) {
  return name.split('.').map(part => _utils.lodash.camelCase(part)).join('.');
}

function pkgNameToKey(pkgName, type) {
  if (pkgName.charAt(0) === '@' && !pkgName.startsWith('@fesjs/')) {
    pkgName = pkgName.split('/')[1];
  }

  return nameToKey(pkgName.replace(RE[type], ''));
}

function pathToObj({
  path,
  type,
  cwd
}) {
  let pkg = null;
  let isPkgPlugin = false;

  const pkgJSONPath = _utils.pkgUp.sync({
    cwd: path
  });

  if (pkgJSONPath) {
    // eslint-disable-next-line
    pkg = require(pkgJSONPath);
    isPkgPlugin = (0, _utils.winPath)((0, _path.join)((0, _path.dirname)(pkgJSONPath), pkg.main || 'index.js')) === (0, _utils.winPath)(path);
  }

  let id;

  if (isPkgPlugin) {
    id = pkg.name;
  } else if ((0, _utils.winPath)(path).startsWith((0, _utils.winPath)(cwd))) {
    id = `./${(0, _utils.winPath)((0, _path.relative)(cwd, path))}`;
  } else if (pkgJSONPath) {
    id = (0, _utils.winPath)((0, _path.join)(pkg.name, (0, _path.relative)((0, _path.dirname)(pkgJSONPath), path)));
  } else {
    id = (0, _utils.winPath)(path);
  }

  id = id.replace('@fesjs/preset-built-in/lib/plugins', '@@');
  id = id.replace(/\.js$/, '');
  const key = isPkgPlugin ? pkgNameToKey(pkg.name, type) : nameToKey((0, _path.basename)(path, (0, _path.extname)(path)));
  return {
    id,
    key,
    path: (0, _utils.winPath)(path),

    apply() {
      // use function to delay require
      try {
        // eslint-disable-next-line
        const ret = require(path); // use the default member for es modules


        return (0, _utils.compatESModuleRequire)(ret);
      } catch (e) {
        throw new Error(`Register ${path} failed, since ${e.message}`);
      }
    },

    defaultConfig: null
  };
}

function resolvePresets(opts) {
  const type = _enums.PluginType.preset;
  const presets = [...getPluginsOrPresets(type, opts)];
  return presets.map(path => pathToObj({
    type,
    path,
    cwd: opts.cwd
  }));
}

function resolvePlugins(opts) {
  const type = _enums.PluginType.plugin;
  const plugins = getPluginsOrPresets(type, opts);
  return plugins.map(path => pathToObj({
    path,
    type,
    cwd: opts.cwd
  }));
}

function isValidPlugin(plugin) {
  return plugin.id && plugin.key && plugin.apply;
}
