// @fesjs/compiler/lib/service/pluginAPI.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _assert = _interopRequireDefault(require("assert"));

var utils = _interopRequireWildcard(require("@fesjs/utils"));

var _pluginUtils = require("./utils/pluginUtils");

var _enums = require("./enums");

var _logger = _interopRequireDefault(require("../logger"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @copy 该文件代码大部分出自 umi，有需要请参考：
 * https://github.com/umijs/umi/tree/master/packages/core
 */
// TODO
// 标准化 logger
class PluginAPI {
  constructor(opts) {
    this.id = opts.id;
    this.key = opts.key;
    this.service = opts.service;
    this.utils = utils;
    this.logger = new _logger.default(`fes:plugin:${this.id || this.key}`);
  } // TODO: reversed keys


  describe({
    id,
    key,
    config,
    enableBy
  } = {}) {
    const {
      plugins
    } = this.service; // this.id and this.key is generated automatically
    // so we need to diff first

    if (id && this.id !== id) {
      if (plugins[id]) {
        const name = plugins[id].isPreset ? 'preset' : 'plugin';
        throw new Error(`api.describe() failed, ${name} ${id} is already registered by ${plugins[id].path}.`);
      }

      plugins[id] = plugins[this.id];
      plugins[id].id = id;
      delete plugins[this.id];
      this.id = id;
    }

    if (key && this.key !== key) {
      this.key = key;
      plugins[this.id].key = key;
    }

    if (config) {
      plugins[this.id].config = config;
    }

    plugins[this.id].enableBy = enableBy || _enums.EnableBy.register;
  }

  register(hook) {
    (0, _assert.default)(hook.key && typeof hook.key === 'string', `api.register() failed, hook.key must supplied and should be string, but got ${hook.key}.`);
    (0, _assert.default)(hook.fn && typeof hook.fn === 'function', `api.register() failed, hook.fn must supplied and should be function, but got ${hook.fn}.`);
    this.service.hooksByPluginId[this.id] = (this.service.hooksByPluginId[this.id] || []).concat(hook);
  }

  registerCommand(commandOption) {
    const {
      command,
      fn
    } = commandOption;
    (0, _assert.default)(!this.service.commands[command], `api.registerCommand() failed, the command ${command} is exists.`);
    (0, _assert.default)(typeof command === 'string', 'api.registerCommand() failed, the command must be String.');
    (0, _assert.default)(typeof fn === 'function', 'api.registerCommand() failed, the fn must be function.');
    this.service.commands[command] = commandOption;
  } // 在 preset 初始化阶段放后面，在插件注册阶段放前面


  registerPlugins(plugins) {
    (0, _assert.default)(this.service.stage === _enums.ServiceStage.initPresets || this.service.stage === _enums.ServiceStage.initPlugins, 'api.registerPlugins() failed, it should only be used in registering stage.');
    (0, _assert.default)(Array.isArray(plugins), 'api.registerPlugins() failed, plugins must be Array.');
    const extraPlugins = plugins.map(plugin => (0, _pluginUtils.isValidPlugin)(plugin) ? plugin : (0, _pluginUtils.pathToObj)({
      type: _enums.PluginType.plugin,
      path: plugin,
      cwd: this.service.cwd
    }));

    if (this.service.stage === _enums.ServiceStage.initPresets) {
      this.service._extraPlugins.push(...extraPlugins);
    } else {
      this.service._extraPlugins.splice(0, 0, ...extraPlugins);
    }
  }

  registerPresets(presets) {
    (0, _assert.default)(this.service.stage === _enums.ServiceStage.initPresets, 'api.registerPresets() failed, it should only used in presets.');
    (0, _assert.default)(Array.isArray(presets), 'api.registerPresets() failed, presets must be Array.');
    const extraPresets = presets.map(preset => (0, _pluginUtils.isValidPlugin)(preset) ? preset : (0, _pluginUtils.pathToObj)({
      type: _enums.PluginType.preset,
      path: preset,
      cwd: this.service.cwd
    })); // 插到最前面，下个 while 循环优先执行

    this.service._extraPresets.splice(0, 0, ...extraPresets);
  }

  registerMethod({
    name,
    fn,
    exitsError = true
  }) {
    if (this.service.pluginMethods[name]) {
      if (exitsError) {
        throw new Error(`api.registerMethod() failed, method ${name} is already exist.`);
      } else {
        return;
      }
    }

    this.service.pluginMethods[name] = fn // 这里不能用 arrow function，this 需指向执行此方法的 PluginAPI
    // 否则 pluginId 会不会，导致不能正确 skip plugin
    || function (hookFn) {
      const hook = {
        key: name,
        ...(utils.lodash.isPlainObject(hookFn) ? hookFn : {
          fn: hookFn
        })
      }; // @ts-ignore

      this.register(hook);
    };
  }

  skipPlugins(pluginIds) {
    pluginIds.forEach(pluginId => {
      this.service.skipPluginIds.add(pluginId);
    });
  }

}

exports.default = PluginAPI;
