// @form-create/component-antdv-frame/dist/index.js
/*!
 * @form-create/component-antdv-frame v3.1.3
 * (c) 2018-2022 xaboy
 * Github https://github.com/xaboy/form-create with frame
 * Released under the MIT License.
 */
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('vue')) :
  typeof define === 'function' && define.amd ? define(['exports', 'vue'], factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.FcFrame = {}, global.Vue));
})(this, (function (exports, vue) { 'use strict';

  function ownKeys(object, enumerableOnly) {
    var keys = Object.keys(object);

    if (Object.getOwnPropertySymbols) {
      var symbols = Object.getOwnPropertySymbols(object);

      if (enumerableOnly) {
        symbols = symbols.filter(function (sym) {
          return Object.getOwnPropertyDescriptor(object, sym).enumerable;
        });
      }

      keys.push.apply(keys, symbols);
    }

    return keys;
  }

  function _objectSpread2(target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i] != null ? arguments[i] : {};

      if (i % 2) {
        ownKeys(Object(source), true).forEach(function (key) {
          _defineProperty(target, key, source[key]);
        });
      } else if (Object.getOwnPropertyDescriptors) {
        Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));
      } else {
        ownKeys(Object(source)).forEach(function (key) {
          Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
        });
      }
    }

    return target;
  }

  function _defineProperty(obj, key, value) {
    if (key in obj) {
      Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
      });
    } else {
      obj[key] = value;
    }

    return obj;
  }

  function toArray(value) {
    return Array.isArray(value) ? value : [null, undefined, ''].indexOf(value) > -1 ? [] : [value];
  }

  // https://github.com/developit/mitt
  function Mitt(all) {
    all = all || new Map();
    var mitt = {
      $on: function $on(type, handler) {
        var handlers = all.get(type);
        var added = handlers && handlers.push(handler);

        if (!added) {
          all.set(type, [handler]);
        }
      },
      $once: function $once(type, handler) {
        handler._once = true;
        mitt.$on(type, handler);
      },
      $off: function $off(type, handler) {
        var handlers = all.get(type);

        if (handlers) {
          handlers.splice(handlers.indexOf(handler) >>> 0, 1);
        }
      },
      $emit: function $emit(type) {
        for (var _len = arguments.length, args = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
          args[_key - 1] = arguments[_key];
        }

        (all.get(type) || []).slice().map(function (handler) {
          if (handler._once) {
            mitt.$off(type, handler);
            delete handler._once;
          }

          handler.apply(void 0, args);
        });
        (all.get('*') || []).slice().map(function (handler) {
          handler(type, args);
        });
      }
    };
    return mitt;
  }

  var script$4 = {
    name: 'CloseCircleOutlined'
  };

  var _hoisted_1$4 = {
    "class": "anticon"
  };

  var _hoisted_2$4 = /*#__PURE__*/vue.createElementVNode("svg", {
    height: "1em",
    width: "1em",
    xmlns: "http://www.w3.org/2000/svg",
    viewBox: "0 0 1024 1024"
  }, [/*#__PURE__*/vue.createElementVNode("path", {
    d: "M685.4 354.8c0-4.4-3.6-8-8-8l-66 .3L512 465.6l-99.3-118.4l-66.1-.3c-4.4 0-8 3.5-8 8c0 1.9.7 3.7 1.9 5.2l130.1 155L340.5 670a8.32 8.32 0 0 0-1.9 5.2c0 4.4 3.6 8 8 8l66.1-.3L512 564.4l99.3 118.4l66 .3c4.4 0 8-3.5 8-8c0-1.9-.7-3.7-1.9-5.2L553.5 515l130.1-155c1.2-1.4 1.8-3.3 1.8-5.2z",
    fill: "currentColor"
  }), /*#__PURE__*/vue.createElementVNode("path", {
    d: "M512 65C264.6 65 64 265.6 64 513s200.6 448 448 448s448-200.6 448-448S759.4 65 512 65zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372s372 166.6 372 372s-166.6 372-372 372z",
    fill: "currentColor"
  })], -1);

  var _hoisted_3$4 = [_hoisted_2$4];
  function render$4(_ctx, _cache, $props, $setup, $data, $options) {
    return vue.openBlock(), vue.createElementBlock("span", _hoisted_1$4, _hoisted_3$4);
  }

  script$4.render = render$4;

  var script$3 = {
    name: 'FolderOutlined'
  };

  var _hoisted_1$3 = {
    "class": "anticon"
  };

  var _hoisted_2$3 = /*#__PURE__*/vue.createElementVNode("svg", {
    height: "1em",
    width: "1em",
    xmlns: "http://www.w3.org/2000/svg",
    viewBox: "0 0 1024 1024"
  }, [/*#__PURE__*/vue.createElementVNode("path", {
    d: "M880 298.4H521L403.7 186.2a8.15 8.15 0 0 0-5.5-2.2H144c-17.7 0-32 14.3-32 32v592c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V330.4c0-17.7-14.3-32-32-32zM840 768H184V256h188.5l119.6 114.4H840V768z",
    fill: "currentColor"
  })], -1);

  var _hoisted_3$3 = [_hoisted_2$3];
  function render$3(_ctx, _cache, $props, $setup, $data, $options) {
    return vue.openBlock(), vue.createElementBlock("span", _hoisted_1$3, _hoisted_3$3);
  }

  script$3.render = render$3;

  var script$2 = {
    name: 'FileOutlined'
  };

  var _hoisted_1$2 = {
    "class": "anticon"
  };

  var _hoisted_2$2 = /*#__PURE__*/vue.createElementVNode("svg", {
    height: "1em",
    width: "1em",
    xmlns: "http://www.w3.org/2000/svg",
    viewBox: "0 0 1024 1024"
  }, [/*#__PURE__*/vue.createElementVNode("path", {
    d: "M854.6 288.6L639.4 73.4c-6-6-14.1-9.4-22.6-9.4H192c-17.7 0-32 14.3-32 32v832c0 17.7 14.3 32 32 32h640c17.7 0 32-14.3 32-32V311.3c0-8.5-3.4-16.7-9.4-22.7zM790.2 326H602V137.8L790.2 326zm1.8 562H232V136h302v216a42 42 0 0 0 42 42h216v494z",
    fill: "currentColor"
  })], -1);

  var _hoisted_3$2 = [_hoisted_2$2];
  function render$2(_ctx, _cache, $props, $setup, $data, $options) {
    return vue.openBlock(), vue.createElementBlock("span", _hoisted_1$2, _hoisted_3$2);
  }

  script$2.render = render$2;

  var script$1 = {
    name: 'DeleteOutlined'
  };

  var _hoisted_1$1 = {
    "class": "anticon"
  };

  var _hoisted_2$1 = /*#__PURE__*/vue.createElementVNode("svg", {
    height: "1em",
    width: "1em",
    xmlns: "http://www.w3.org/2000/svg",
    viewBox: "0 0 1024 1024"
  }, [/*#__PURE__*/vue.createElementVNode("path", {
    d: "M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z",
    fill: "currentColor"
  })], -1);

  var _hoisted_3$1 = [_hoisted_2$1];
  function render$1(_ctx, _cache, $props, $setup, $data, $options) {
    return vue.openBlock(), vue.createElementBlock("span", _hoisted_1$1, _hoisted_3$1);
  }

  script$1.render = render$1;

  var script = {
    name: 'EyeOutlined'
  };

  var _hoisted_1 = {
    "class": "anticon"
  };

  var _hoisted_2 = /*#__PURE__*/vue.createElementVNode("svg", {
    height: "1em",
    width: "1em",
    xmlns: "http://www.w3.org/2000/svg",
    viewBox: "0 0 1024 1024"
  }, [/*#__PURE__*/vue.createElementVNode("path", {
    d: "M942.2 486.2C847.4 286.5 704.1 186 512 186c-192.2 0-335.4 100.5-430.2 300.3a60.3 60.3 0 0 0 0 51.5C176.6 737.5 319.9 838 512 838c192.2 0 335.4-100.5 430.2-300.3c7.7-16.2 7.7-35 0-51.5zM512 766c-161.3 0-279.4-81.8-362.7-254C232.6 339.8 350.7 258 512 258c161.3 0 279.4 81.8 362.7 254C791.5 684.2 673.4 766 512 766zm-4-430c-97.2 0-176 78.8-176 176s78.8 176 176 176s176-78.8 176-176s-78.8-176-176-176zm0 288c-61.9 0-112-50.1-112-112s50.1-112 112-112s112 50.1 112 112s-50.1 112-112 112z",
    fill: "currentColor"
  })], -1);

  var _hoisted_3 = [_hoisted_2];
  function render(_ctx, _cache, $props, $setup, $data, $options) {
    return vue.openBlock(), vue.createElementBlock("span", _hoisted_1, _hoisted_3);
  }

  script.render = render;

  function styleInject(css, ref) {
    if ( ref === void 0 ) ref = {};
    var insertAt = ref.insertAt;

    if (!css || typeof document === 'undefined') { return; }

    var head = document.head || document.getElementsByTagName('head')[0];
    var style = document.createElement('style');
    style.type = 'text/css';

    if (insertAt === 'top') {
      if (head.firstChild) {
        head.insertBefore(style, head.firstChild);
      } else {
        head.appendChild(style);
      }
    } else {
      head.appendChild(style);
    }

    if (style.styleSheet) {
      style.styleSheet.cssText = css;
    } else {
      style.appendChild(document.createTextNode(css));
    }
  }

  var css_248z = "._fc-frame ._fc-files img{display:inline-block;height:100%;vertical-align:top;width:100%}._fc-frame ._fc-upload-cover{background:rgba(0,0,0,.6);bottom:0;left:0;opacity:0;position:absolute;right:0;top:0;transition:opacity .3s}._fc-frame ._fc-upload-cover ._fc-frame-icon{color:#fff;font-size:16px;margin:0 2px}._fc-frame ._fc-files:hover ._fc-upload-cover{opacity:1}._fc-frame .anticon{font-size:16px;width:16px}._fc-frame .ant-upload{display:block}._fc-frame ._fc-frame-icon,._fc-frame ._fc-upload-btn{cursor:pointer}._fc-files,._fc-frame ._fc-upload-btn{background:#fff;border:1px solid #c0ccda;border-radius:4px;box-shadow:2px 2px 5px rgba(0,0,0,.1);box-sizing:border-box;display:inline-block;height:104px;line-height:104px;margin-right:4px;overflow:hidden;position:relative;text-align:center;width:104px}";
  styleInject(css_248z);

  function _isSlot(s) {
    return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !vue.isVNode(s);
  }

  var NAME = 'fcFrame';
  var Frame = vue.defineComponent({
    name: NAME,
    props: {
      type: {
        type: String,
        "default": 'input'
      },
      field: String,
      helper: {
        type: Boolean,
        "default": true
      },
      disabled: {
        type: Boolean,
        "default": false
      },
      src: {
        type: String,
        required: true
      },
      icon: {
        type: String,
        "default": 'FolderOutlined'
      },
      width: {
        type: [Number, String],
        "default": 500
      },
      height: {
        type: String,
        "default": '370px'
      },
      maxLength: {
        type: Number,
        "default": 0
      },
      okBtnText: {
        type: String,
        "default": '确定'
      },
      closeBtnText: {
        type: String,
        "default": '关闭'
      },
      modalTitle: String,
      handleIcon: {
        type: [String, Boolean],
        "default": undefined
      },
      title: String,
      allowRemove: {
        type: Boolean,
        "default": true
      },
      onOpen: {
        type: Function,
        "default": function _default() {}
      },
      onOk: {
        type: Function,
        "default": function _default() {}
      },
      onCancel: {
        type: Function,
        "default": function _default() {}
      },
      onLoad: {
        type: Function,
        "default": function _default() {}
      },
      onBeforeRemove: {
        type: Function,
        "default": function _default() {}
      },
      onRemove: {
        type: Function,
        "default": function _default() {}
      },
      onHandle: Function,
      modal: {
        type: Object,
        "default": function _default() {
          return {};
        }
      },
      srcKey: [String, Number],
      modelValue: [Array, String, Number, Object],
      previewMask: undefined,
      footer: {
        type: Boolean,
        "default": true
      },
      reload: {
        type: Boolean,
        "default": true
      },
      closeBtn: {
        type: Boolean,
        "default": true
      },
      okBtn: {
        type: Boolean,
        "default": true
      },
      formCreateInject: Object
    },
    emits: ['update:modelValue', 'change'],
    components: {
      FolderOutlined: script$3,
      EyeOutlined: script
    },
    data: function data() {
      return {
        fileList: toArray(this.modelValue),
        previewVisible: false,
        frameVisible: false,
        previewImage: '',
        bus: new Mitt()
      };
    },
    watch: {
      modelValue: function modelValue(n) {
        this.fileList = toArray(n);
      }
    },
    methods: {
      close: function close() {
        this.closeModal(true);
      },
      closeModal: function closeModal(close) {
        this.bus.$emit(close ? '$close' : '$ok');

        if (this.reload) {
          this.bus.$off('$ok');
          this.bus.$off('$close');
        }

        this.frameVisible = false;
      },
      handleCancel: function handleCancel() {
        this.previewVisible = false;
      },
      showModal: function showModal() {
        if (this.disabled || false === this.onOpen()) {
          return;
        }

        this.frameVisible = true;
      },
      input: function input() {
        var n = this.fileList;
        var val = this.maxLength === 1 ? n[0] || '' : n;
        this.$emit('update:modelValue', val);
        this.$emit('change', val);
      },
      makeInput: function makeInput() {
        var _this = this;

        var Type = vue.resolveComponent(this.icon);
        var slots = {
          addonAfter: function addonAfter() {
            return vue.createVNode(Type, {
              "class": "_fc-frame-icon",
              "onClick": _this.showModal
            }, null);
          }
        };

        if (this.fileList.length) {
          slots.suffix = function () {
            return vue.createVNode(script$4, {
              "class": "_fc-frame-icon",
              "onClick": function onClick() {
                _this.fileList = [];

                _this.input();
              }
            }, null);
          };
        }

        return vue.createVNode(vue.resolveComponent("AInput"), {
          "readonly": true,
          "value": this.fileList.map(function (v) {
            return _this.getSrc(v);
          }).toString(),
          "key": 1
        }, slots);
      },
      makeGroup: function makeGroup(children) {
        if (!this.maxLength || this.fileList.length < this.maxLength) children.push(this.makeBtn());
        return vue.createVNode("div", {
          "key": 2
        }, [children]);
      },
      makeItem: function makeItem(index, children) {
        return vue.createVNode("div", {
          "class": "_fc-files",
          "key": '3' + index
        }, [children]);
      },
      valid: function valid(f) {
        var field = this.formCreateInject.field || this.field;

        if (field && f !== field) {
          throw new Error('[frame]无效的字段值');
        }
      },
      makeIcons: function makeIcons(val, index) {
        if (this.handleIcon !== false || this.allowRemove === true) {
          var icons = [];
          if (this.type !== 'file' && this.handleIcon !== false || this.type === 'file' && this.handleIcon) icons.push(this.makeHandleIcon(val, index));
          if (this.allowRemove) icons.push(this.makeRemoveIcon(val, index));
          return vue.createVNode("div", {
            "class": "_fc-upload-cover",
            "key": 4
          }, [icons]);
        }
      },
      makeHandleIcon: function makeHandleIcon(val, index) {
        var _this2 = this;

        var Type = vue.resolveComponent(this.handleIcon === true || this.handleIcon === undefined ? 'EyeOutlined' : this.handleIcon);
        return vue.createVNode(Type, {
          "class": "_fc-frame-icon",
          "onClick": function onClick() {
            return _this2.handleClick(val);
          },
          "key": '5' + index
        }, null);
      },
      makeRemoveIcon: function makeRemoveIcon(val, index) {
        var _this3 = this;

        return vue.createVNode(script$1, {
          "class": "_fc-frame-icon",
          "onClick": function onClick() {
            return _this3.handleRemove(val);
          },
          "key": '6' + index
        }, null);
      },
      makeFiles: function makeFiles() {
        var _this4 = this;

        return this.makeGroup(this.fileList.map(function (src, index) {
          return _this4.makeItem(index, [vue.createVNode(script$2, {
            "class": "_fc-frame-icon",
            "onClick": function onClick() {
              return _this4.handleClick(src);
            }
          }, null), _this4.makeIcons(src, index)]);
        }));
      },
      makeImages: function makeImages() {
        var _this5 = this;

        return this.makeGroup(this.fileList.map(function (src, index) {
          return _this5.makeItem(index, [vue.createVNode("img", {
            "src": _this5.getSrc(src)
          }, null), _this5.makeIcons(src, index)]);
        }));
      },
      makeBtn: function makeBtn() {
        var _this6 = this;

        var Type = vue.resolveComponent(this.icon);
        return vue.createVNode("div", {
          "class": "_fc-upload-btn",
          "onClick": function onClick() {
            return _this6.showModal();
          },
          "key": 7
        }, [vue.createVNode(Type, {
          "class": "_fc-frame-icon"
        }, null)]);
      },
      handleClick: function handleClick(src) {
        if (this.onHandle) {
          return this.onHandle(src);
        } else {
          this.previewImage = this.getSrc(src);
          this.previewVisible = true;
        }
      },
      handleRemove: function handleRemove(src) {
        if (this.disabled) return;

        if (false !== this.onBeforeRemove(src)) {
          this.fileList.splice(this.fileList.indexOf(src), 1);
          this.input();
          this.onRemove(src);
        }
      },
      getSrc: function getSrc(src) {
        return !this.srcKey ? src : src[this.srcKey];
      },
      frameLoad: function frameLoad(iframe) {
        var _this7 = this;

        this.onLoad(iframe);

        try {
          if (this.helper === true) {
            iframe['form_create_helper'] = {
              api: this.formCreateInject.api,
              close: function close(field) {
                _this7.valid(field);

                _this7.closeModal();
              },
              set: function set(field, value) {
                _this7.valid(field);

                if (!_this7.disabled) _this7.$emit('update:modelValue', value);
              },
              get: function get(field) {
                _this7.valid(field);

                return _this7.modelValue;
              },
              onOk: function onOk(fn) {
                return _this7.bus.$on('$ok', fn);
              },
              onClose: function onClose(fn) {
                return _this7.bus.$on('$close', fn);
              }
            };
          }
        } catch (e) {
          console.error(e);
        }
      },
      makeFooter: function makeFooter() {
        var _this8 = this;

        var _this$$props = this.$props,
            okBtnText = _this$$props.okBtnText,
            closeBtnText = _this$$props.closeBtnText,
            closeBtn = _this$$props.closeBtn,
            okBtn = _this$$props.okBtn,
            footer = _this$$props.footer;
        var node = [];
        if (!footer) return node;
        if (closeBtn) node.push(vue.createVNode(vue.resolveComponent("AButton"), {
          "onClick": function onClick() {
            return _this8.onCancel() !== false && _this8.closeModal(true);
          }
        }, _isSlot(closeBtnText) ? closeBtnText : {
          "default": function _default() {
            return [closeBtnText];
          }
        }));
        if (okBtn) node.push(vue.createVNode(vue.resolveComponent("AButton"), {
          "type": "primary",
          "onClick": function onClick() {
            return _this8.onOk() !== false && _this8.closeModal();
          }
        }, _isSlot(okBtnText) ? okBtnText : {
          "default": function _default() {
            return [okBtnText];
          }
        }));
        return node;
      }
    },
    render: function render() {
      var _this9 = this;

      var type = this.type;
      var Node;
      if (type === 'input') Node = this.makeInput();else if (type === 'image') Node = this.makeImages();else Node = this.makeFiles();
      var _this$$props2 = this.$props,
          _this$$props2$width = _this$$props2.width,
          width = _this$$props2$width === void 0 ? '30%' : _this$$props2$width,
          height = _this$$props2.height,
          src = _this$$props2.src,
          title = _this$$props2.title,
          modalTitle = _this$$props2.modalTitle;
      vue.nextTick(function () {
        if (_this9.$refs.frame) {
          _this9.frameLoad(_this9.$refs.frame.contentWindow || {});
        }
      });
      return vue.createVNode("div", {
        "class": "_fc-frame"
      }, [Node, vue.createVNode(vue.resolveComponent("aModal"), {
        "mask": this.previewMask,
        "title": modalTitle,
        "visible": this.previewVisible,
        "onCancel": function onCancel() {
          return _this9.previewVisible = false;
        },
        "footer": null
      }, {
        "default": function _default() {
          return [vue.createVNode("img", {
            "style": "width: 100%",
            "src": _this9.previewImage
          }, null)];
        }
      }), vue.createVNode(vue.resolveComponent("aModal"), vue.mergeProps(_objectSpread2({
        width: width,
        title: title
      }, this.modal), {
        "visible": this.frameVisible,
        "onCancel": function onCancel() {
          return _this9.closeModal(true);
        }
      }), {
        "default": function _default() {
          return [_this9.frameVisible || !_this9.reload ? vue.createVNode("iframe", {
            "ref": "frame",
            "src": src,
            "frameborder": "0",
            "style": {
              height: height,
              'border': '0 none',
              'width': '100%'
            }
          }, null) : null];
        },
        footer: function footer() {
          return _this9.makeFooter();
        }
      })]);
    },
    beforeMount: function beforeMount() {
      var _this$formCreateInjec = this.formCreateInject,
          name = _this$formCreateInjec.name,
          field = _this$formCreateInjec.field,
          api = _this$formCreateInjec.api;
      name && api.on('fc:closeModal:' + name, this.close);
      field && api.on('fc:closeModal:' + field, this.close);
    },
    beforeUnmount: function beforeUnmount() {
      var _this$formCreateInjec2 = this.formCreateInject,
          name = _this$formCreateInjec2.name,
          field = _this$formCreateInjec2.field,
          api = _this$formCreateInjec2.api;
      name && api.off('fc:closeModal:' + name, this.close);
      field && api.off('fc:closeModal:' + field, this.close);
    }
  });

  exports["default"] = Frame;

  Object.defineProperty(exports, '__esModule', { value: true });

}));
