// @fesjs/preset-built-in/lib/plugins/generateFiles/core/plugin/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _default;

var _fs = require("fs");

var _path = require("path");

var _utils = require("@fesjs/utils");

var _constants = require("../../../../utils/constants");

var _getAppEntryPath = require("../../../../utils/getAppEntryPath");

function _default(api) {
  const {
    paths,
    utils: {
      Mustache
    }
  } = api;
  const absoluteFilePath = 'core/plugin.js';
  api.onGenerateFiles(async () => {
    const validKeys = await api.applyPlugins({
      key: 'addRuntimePluginKey',
      type: api.ApplyPluginsType.add,
      initialValue: [// 初始化数据
      'beforeRender', // modify渲染工具
      'modifyClientRenderOpts', 'rootContainer', // app生成时触发
      'onAppCreated', // 渲染
      'render', // 修改路由
      'patchRoutes', // 修改history
      'modifyCreateHistory', // 生成router时触发
      'onRouterCreated']
    });
    const plugins = await api.applyPlugins({
      key: 'addRuntimePlugin',
      type: api.ApplyPluginsType.add,
      initialValue: [(0, _getAppEntryPath.getAppPath)(paths.absSrcPath)].filter(Boolean)
    });
    api.writeTmpFile({
      path: absoluteFilePath,
      content: Mustache.render((0, _fs.readFileSync)((0, _path.join)(__dirname, 'plugin.tpl'), 'utf-8'), {
        validKeys,
        runtimePath: _constants.runtimePath
      })
    });
    api.writeTmpFile({
      path: 'core/pluginRegister.js',
      content: Mustache.render((0, _fs.readFileSync)((0, _path.join)(__dirname, 'pluginRegister.tpl'), 'utf-8'), {
        plugins: plugins.map((plugin, index) => ({
          index,
          path: (0, _utils.winPath)(plugin)
        }))
      })
    });
  });
  api.addCoreExports(() => ({
    specifiers: ['plugin'],
    source: absoluteFilePath
  }));
}
