// @fesjs/preset-built-in/lib/utils/generateFiles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _utils = require("@fesjs/utils");

var _getAppEntryPath = require("./getAppEntryPath");

var _default = async ({
  api,
  watch
}) => {
  const {
    paths
  } = api;

  async function generate() {
    api.logger.debug('generate files');
    await api.applyPlugins({
      key: 'onGenerateFiles',
      type: api.ApplyPluginsType.event
    });
  }

  let watchers = [];
  await generate();

  function unwatch() {
    watchers.forEach(watcher => {
      watcher.close();
    });
    watchers = [];
  }

  function createWatcher(path) {
    const watcher = _utils.chokidar.watch(path, {
      // ignore .dotfiles and _mock.js
      ignored: /(^|[/\\])(_mock.js$|\..)/,
      ignoreInitial: true
    });

    watcher.on('all', _utils.lodash.throttle(async () => {
      await generate();
    }, 100));
    watchers.push(watcher);
  }

  if (watch) {
    const watcherPaths = await api.applyPlugins({
      key: 'addTmpGenerateWatcherPaths',
      type: api.ApplyPluginsType.add,
      initialValue: [paths.absPagesPath, (0, _getAppEntryPath.getAppPath)(paths.absSrcPath)]
    });

    _utils.lodash.uniq(watcherPaths.map(p => (0, _utils.winPath)(p))).forEach(p => {
      createWatcher(p);
    });
  }

  return unwatch;
};

exports.default = _default;
