// @fesjs/plugin-enums/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _fs = require("fs");

var _path = require("path");

const namespace = 'plugin-enums';

var _default = api => {
  const {
    utils: {
      Mustache
    }
  } = api;
  api.describe({
    key: 'enums',
    config: {
      schema(joi) {
        return joi.object();
      },

      onChange: api.ConfigChangeType.regenerateTmpFiles
    }
  });
  const absoluteFilePath = (0, _path.join)(namespace, 'core.js');
  api.onGenerateFiles(() => {
    // 文件写出
    const enums = api.config.enums || {};
    api.writeTmpFile({
      path: absoluteFilePath,
      content: Mustache.render((0, _fs.readFileSync)((0, _path.join)(__dirname, 'runtime/core.tpl'), 'utf-8'), {
        REPLACE_ENUMS: JSON.stringify(enums)
      })
    });
    api.copyTmpFiles({
      namespace,
      path: (0, _path.join)(__dirname, 'runtime'),
      ignore: ['.tpl']
    });
  });
  api.addPluginExports(() => [{
    specifiers: ['enums'],
    source: absoluteFilePath
  }]);
};

exports.default = _default;
