// @fesjs/plugin-layout/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _fs = require("fs");

var _path = require("path");

var _utils = require("@fesjs/utils");

const namespace = 'plugin-layout';

var _default = api => {
  const {
    utils: {
      Mustache
    }
  } = api;

  const helper = require('./node/helper');

  api.describe({
    key: 'layout',
    config: {
      schema(joi) {
        return joi.object();
      },

      onChange: api.ConfigChangeType.regenerateTmpFiles
    }
  });
  api.addRuntimePluginKey(() => 'layout');
  const absFilePath = (0, _path.join)(namespace, 'index.js');
  const absRuntimeFilePath = (0, _path.join)(namespace, 'runtime.js');
  api.onGenerateFiles(async () => {
    const {
      name
    } = api.pkg;
    const HAS_LOCALE = api.hasPlugins(['@fesjs/plugin-locale']); // .fes配置

    const userConfig = {
      title: name,
      footer: 'Created by Fes.js',
      ...(api.config.layout || {})
    }; // 路由信息

    const routes = await api.getRoutes(); // 把路由的meta合并到menu配置中

    userConfig.menus = helper.fillMenuByRoute(userConfig.menus, routes);
    const icons = helper.getIconsFromMenu(userConfig.menus);
    const iconsString = icons.map(iconName => `import ${iconName} from '@ant-design/icons-vue/es/icons/${iconName}'`);
    api.writeTmpFile({
      path: (0, _path.join)(namespace, 'icons.js'),
      content: `
        ${iconsString.join(';\n')}
        export default {
            ${icons.join(',\n')}
        }`
    });
    api.writeTmpFile({
      path: absFilePath,
      content: Mustache.render((0, _fs.readFileSync)((0, _path.join)(__dirname, 'runtime/index.tpl'), 'utf-8'), {
        REPLACE_USER_CONFIG: JSON.stringify(userConfig),
        HAS_LOCALE
      })
    });
    api.copyTmpFiles({
      namespace,
      path: (0, _path.join)(__dirname, 'runtime'),
      ignore: ['.tpl']
    });
  });
  api.addRuntimePlugin(() => `@@/${absRuntimeFilePath}`); // 把BaseLayout插入到路由配置中，作为根路由

  api.modifyRoutes(routes => [{
    path: '/',
    component: (0, _utils.winPath)((0, _path.join)(api.paths.absTmpPath || '', absFilePath)),
    children: routes
  }]);
};

exports.default = _default;
