// @webank/eslint-config-webank/angular.js
module.exports = {
    extends: ['./index.js'],
    globals: {
        angular: true,
        wa: true,
        WaterMark: true,
        _: true,
        softphone: true,
        WeakMap: true,
        pinyinUtil: true,
        TRTC: true,
        iframe: true,
        portalBridge: true,
        uParse: true,
        filterXSS: true,
        RtcClient: true
    },
    env: {
        browser: true,
        jquery: true,
        es6: true
    },
    plugins: ['angular'],
    rules: {
        /* error */
        'angular/avoid-scope-typos': 2,
        'angular/module-getter': 2,
        'angular/module-setter': 2,
        'angular/no-private-call': 2,
        /* best practices */
        'angular/component-name': 1,
        'angular/constant-name': 1,
        'angular/controller-as-route': 0,
        'angular/controller-as-vm': 0,
        'angular/controller-as': 0,
        'angular/controller-name': 0,
        'angular/directive-name': 1,
        'angular/directive-restrict': 1,
        'angular/document-service': 0,
        'angular/factory-name': 1,
        'angular/file-name': 0,
        'angular/filter-name': 1,
        'angular/function-type': 0,
        'angular/interval-service': 1,
        'angular/module-name': 1,
        'angular/no-run-logic': 0,
        'angular/no-service-method': 0,
        'angular/provider-name': 1,
        'angular/service-name': 1,
        'angular/timeout-service': 0,
        'angular/value-name': 1,
        'angular/window-service': 0
    }
};
