// @webank/eslint-config-webank/index.js
/**
 * eslint 规则
 *
 * @fixable 表示此配置支持 --fix
 * @off 表示此配置被关闭了，后面说明关闭的原因
 */
module.exports = {
    parser: '@babel/eslint-parser',
    extends: [
        'eslint:recommended',
        './rules/best-practices',
        './rules/errors',
        './rules/style',
        './rules/variables',
        './rules/es6',
        './rules/imports'
    ],
    parserOptions: {
        requireConfigFile: false,
        ecmaVersion: 2021,
        sourceType: 'module',
        ecmaFeatures: {
            experimentalObjectRestSpread: true,
            legacyDecorators: true,
            impliedStrict: true,
            jsx: true,
            moduels: true
        },
        babelOptions: {
            plugins: [
                '@babel/plugin-syntax-jsx'
            ]
        }
    },
    plugins: [
        'html'
    ],
    env: {
        browser: true,
        node: true,
        commonjs: true,
        es6: true
    },
    // 以当前目录为根目录，不再向上查找 .eslintrc.js
    root: true,
    rules: {
        // @fixable 禁止使用 ++ --
        // @off 没有必要
        'prefer-destructuring': 'off'
    }
};
