// @fesjs/plugin-request/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _compiler = require("@fesjs/compiler");

var _fs = require("fs");

var _path = require("path");

var _utils = require("@fesjs/utils");

const logger = new _compiler.Logger('fes:plugin-request');

var _default = api => {
  api.addRuntimePluginKey(() => 'request'); // 配置

  api.describe({
    key: 'request',
    config: {
      schema(joi) {
        return joi.object({
          dataField: joi.string().pattern(/^[a-zA-Z]*$/).allow(''),
          base: joi.string().allow('')
        });
      },

      default: {
        base: '',
        dataField: ''
      }
    }
  });
  const namespace = 'plugin-request';
  const absoluteFilePath = `${namespace}/request.js`;
  const requestTemplate = (0, _fs.readFileSync)((0, _path.join)(__dirname, 'template', 'request.js'), 'utf-8');
  api.onGenerateFiles(() => {
    // 文件写出
    const {
      dataField = '',
      base = ''
    } = api.config.request;

    if (base) {
      // DEPRECATED
      logger.warn('[DEPRECATED]: reqeust base 即将废弃，建议使用 axios baseURL代替：https://github.com/axios/axios');
    }

    api.writeTmpFile({
      path: absoluteFilePath,
      content: requestTemplate.replace('REPLACE_DATA_FIELD', JSON.stringify(dataField)).replace('REPLACE_BASE', base || '').replace('AXIOS_PATH', (0, _utils.resolvePkg)('axios'))
    });
  });
  let generatedOnce = false;
  api.onGenerateFiles(() => {
    if (generatedOnce) return;
    generatedOnce = true;
    api.copyTmpFiles({
      namespace,
      path: (0, _path.join)(__dirname, 'template'),
      ignore: ['request.js']
    });
  });
  api.addPluginExports(() => [{
    exportAll: true,
    source: absoluteFilePath
  }]);
};

exports.default = _default;
