// @fesjs/compiler/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Config", {
  enumerable: true,
  get: function () {
    return _config.default;
  }
});
Object.defineProperty(exports, "Logger", {
  enumerable: true,
  get: function () {
    return _logger.default;
  }
});
Object.defineProperty(exports, "PluginAPI", {
  enumerable: true,
  get: function () {
    return _pluginAPI.default;
  }
});
Object.defineProperty(exports, "PluginType", {
  enumerable: true,
  get: function () {
    return _enums.PluginType;
  }
});
Object.defineProperty(exports, "Service", {
  enumerable: true,
  get: function () {
    return _service.default;
  }
});
Object.defineProperty(exports, "isPluginOrPreset", {
  enumerable: true,
  get: function () {
    return _pluginUtils.isPluginOrPreset;
  }
});

var _config = _interopRequireDefault(require("./config"));

var _logger = _interopRequireDefault(require("./logger"));

var _service = _interopRequireDefault(require("./service"));

var _pluginAPI = _interopRequireDefault(require("./service/pluginAPI"));

var _enums = require("./service/enums");

var _pluginUtils = require("./service/utils/pluginUtils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
