// @fesjs/utils/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Generator", {
  enumerable: true,
  get: function () {
    return _Generator.default;
  }
});
Object.defineProperty(exports, "Mustache", {
  enumerable: true,
  get: function () {
    return _mustache.default;
  }
});
Object.defineProperty(exports, "chalk", {
  enumerable: true,
  get: function () {
    return _chalk.default;
  }
});
exports.chokidar = void 0;
Object.defineProperty(exports, "cleanRequireCache", {
  enumerable: true,
  get: function () {
    return _cleanRequireCache.default;
  }
});
Object.defineProperty(exports, "compatESModuleRequire", {
  enumerable: true,
  get: function () {
    return _compatESModuleRequire.default;
  }
});
Object.defineProperty(exports, "createDebug", {
  enumerable: true,
  get: function () {
    return _debug.default;
  }
});
Object.defineProperty(exports, "deepmerge", {
  enumerable: true,
  get: function () {
    return _deepmerge.default;
  }
});
Object.defineProperty(exports, "delay", {
  enumerable: true,
  get: function () {
    return _delay.default;
  }
});
Object.defineProperty(exports, "generator", {
  enumerable: true,
  get: function () {
    return _generator.default;
  }
});
Object.defineProperty(exports, "glob", {
  enumerable: true,
  get: function () {
    return _glob.default;
  }
});
Object.defineProperty(exports, "lodash", {
  enumerable: true,
  get: function () {
    return _lodash.default;
  }
});
Object.defineProperty(exports, "mergeConfig", {
  enumerable: true,
  get: function () {
    return _mergeConfig.default;
  }
});
Object.defineProperty(exports, "mkdirp", {
  enumerable: true,
  get: function () {
    return _mkdirp.default;
  }
});
Object.defineProperty(exports, "parseRequireDeps", {
  enumerable: true,
  get: function () {
    return _parseRequireDeps.default;
  }
});
exports.parser = void 0;
Object.defineProperty(exports, "pkgUp", {
  enumerable: true,
  get: function () {
    return _pkgUp.default;
  }
});
Object.defineProperty(exports, "portfinder", {
  enumerable: true,
  get: function () {
    return _portfinder.default;
  }
});
Object.defineProperty(exports, "resolve", {
  enumerable: true,
  get: function () {
    return _resolve.default;
  }
});
Object.defineProperty(exports, "resolvePkg", {
  enumerable: true,
  get: function () {
    return _resolvePkg.default;
  }
});
Object.defineProperty(exports, "rimraf", {
  enumerable: true,
  get: function () {
    return _rimraf.default;
  }
});
Object.defineProperty(exports, "semver", {
  enumerable: true,
  get: function () {
    return _semver.default;
  }
});
Object.defineProperty(exports, "traverse", {
  enumerable: true,
  get: function () {
    return _traverse.default;
  }
});
Object.defineProperty(exports, "winPath", {
  enumerable: true,
  get: function () {
    return _winPath.default;
  }
});
Object.defineProperty(exports, "yParser", {
  enumerable: true,
  get: function () {
    return _yargsParser.default;
  }
});

var _chalk = _interopRequireDefault(require("chalk"));

var _yargsParser = _interopRequireDefault(require("yargs-parser"));

var _lodash = _interopRequireDefault(require("lodash"));

var chokidar = _interopRequireWildcard(require("chokidar"));

exports.chokidar = chokidar;

var _semver = _interopRequireDefault(require("semver"));

var _deepmerge = _interopRequireDefault(require("deepmerge"));

var _glob = _interopRequireDefault(require("glob"));

var _debug = _interopRequireDefault(require("debug"));

var parser = _interopRequireWildcard(require("@babel/parser"));

exports.parser = parser;

var _traverse = _interopRequireDefault(require("@babel/traverse"));

var _generator = _interopRequireDefault(require("@babel/generator"));

var _rimraf = _interopRequireDefault(require("rimraf"));

var _mkdirp = _interopRequireDefault(require("mkdirp"));

var _pkgUp = _interopRequireDefault(require("pkg-up"));

var _portfinder = _interopRequireDefault(require("portfinder"));

var _resolve = _interopRequireDefault(require("resolve"));

var _mustache = _interopRequireDefault(require("mustache"));

var _Generator = _interopRequireDefault(require("./Generator"));

var _winPath = _interopRequireDefault(require("./winPath"));

var _delay = _interopRequireDefault(require("./delay"));

var _resolvePkg = _interopRequireDefault(require("./resolvePkg"));

var _compatESModuleRequire = _interopRequireDefault(require("./compatESModuleRequire"));

var _cleanRequireCache = _interopRequireDefault(require("./cleanRequireCache"));

var _parseRequireDeps = _interopRequireDefault(require("./parseRequireDeps"));

var _mergeConfig = _interopRequireDefault(require("./mergeConfig"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
